collect_stata/write_json.py

```diff
@@ -0,0 +1,308 @@
+"""write_json.py"""
+__author__ = "Marius Pahl"
+
+import json
+from collections import Counter, OrderedDict
+
+import numpy as np
+import pandas as pd
+
+
+def sorting_dataframe(values, labels, missings, frequencies):
+    """Function to sort values and labels and return sorted dict"""
+    dataframe = pd.DataFrame(
+        {
+            "values": values,
+            "labels": labels,
+            "missings": missings,
+            "frequencies": frequencies,
+        }
+    )
+    dataframe["labels"] = dataframe["labels"].astype(str)
+    dataframe["values"] = pd.to_numeric(dataframe["values"])
+    dataframe.sort_values(by="values", inplace=True)
+    return dataframe.to_dict("list")
+
+
+def uni_cat(elem, file_csv):
+    """Generate dict with frequencies and labels for categorical variables
+
+    Input:
+    elem: dict
+    file_csv: pandas DataFrame
+
+    Output:
+    cat_dict: dict
+    """
+
+    frequencies = []
+    values = []
+    missings = []
+    labels = []
+
+    value_count = file_csv[elem["name"]].value_counts()
+    for value in elem["values"]:
+        try:
+            frequencies.append(int(value_count[value["value"]]))
+        except KeyError:
+            frequencies.append(0)
+        labels.append(value["label"])
+
+        var_value = value["value"]
+
+        if int(value["value"]) >= 0:
+            missings.append(False)
+            values.append(var_value)
+        else:
+            missings.append(True)
+            values.append(var_value)
+
+    cat_dict = sorting_dataframe(values, labels, missings, frequencies)
+
+    return cat_dict
+
+
+def uni_string():
+    """Generate dict with frequencies for nominal variables
+
+    Output:
+    string_dict: OrderedDict
+    """
+
+    string_dict = OrderedDict()
+
+    string_dict["frequencies"] = []
+    string_dict["labels"] = []
+    string_dict["missings"] = []
+    string_dict["values"] = []
+
+    return string_dict
+
+
+def uni_number():
+    """Generate dict with frequencies for numerical variables
+
+    Output:
+    OrderedDict
+    """
+
+    return OrderedDict(frequencies=[], labels=[], labels_de=[], missings=[], values=[])
+
+
+def stats_cat(elem, file_csv):
+    """Generate dict with statistics for categorical variables
+
+    Input:
+    elem: dict
+    file_csv: pandas DataFrame
+
+    Output:
+    dict
+    """
+
+    total = int(file_csv[elem["name"]].size)
+    invalid = int(file_csv[elem["name"]].isnull().sum()) + int(
+        sum(n < 0 for n in file_csv[elem["name"]])
+    )
+    valid = total - invalid
+
+    return {"valid": valid, "invalid": invalid}
+
+
+def stats_string(elem, file_csv):
+    """Generate dict with statistics for nominal variables
+
+    Input:
+    elem: dict
+    file_csv: pandas DataFrame
+
+    Output:
+    dict
+    """
+
+    valid = int(file_csv[elem["name"]].value_counts().sum())
+    invalid = int(file_csv[elem["name"]].isnull().sum())
+    for i in file_csv[elem["name"]]:
+        if i in ("", "."):
```

Comment:
What happens here?

Reply:
#18

Reply:
String Missings like "" and "." also count as invalid.
I implemented:
frequencies = Counter(file_csv[elem["name"]])
    string_missings = frequencies[""] + frequencies["."]
So it might be clearer now.